File: config/initializers/stripe_fees.rb
# frozen_string_literal: true

# Single source of truth for the card-processing fee added to member dues
# payments. The member's full `amount` should land on their dues, so the fee is
# grossed up on top: total = amount / (1 - RATE) + FLAT.
#
# Mirrored in app/javascript/utilities/stripe_fees.ts — keep the two in sync.
module StripeFees
  RATE = 0.03        # processor markup (approximates Stripe's 2.9% + rounding headroom)
  FLAT_CENTS = 30    # per-transaction flat fee

  module_function

  # Total to charge the card, in integer cents, so that `amount_cents` nets to dues.
  def total_cents(amount_cents)
    ((amount_cents / (1 - RATE)) + FLAT_CENTS).round
  end

  # The fee portion, in integer cents.
  def fee_cents(amount_cents)
    total_cents(amount_cents) - amount_cents
  end
end

File: app/javascript/utilities/stripe_fees.ts
// Card-processing fee for member dues payments. The member's full `amount`
// lands on their dues; the fee is grossed up on top.
//
// Mirrors config/initializers/stripe_fees.rb — keep the two in sync.

export const STRIPE_FEE_RATE = 0.03
export const STRIPE_FEE_FLAT_CENTS = 30

/** Total to charge the card, in integer cents, so `amountCents` nets to dues. */
export const totalCents = (amountCents: number): number =>
  Math.round(amountCents / (1 - STRIPE_FEE_RATE) + STRIPE_FEE_FLAT_CENTS)

/** The fee portion, in integer cents. */
export const feeCents = (amountCents: number): number =>
  totalCents(amountCents) - amountCents
